File: Basic/Java/Day-10/NextGreater.java
import java.util.Deque;
import java.util.ArrayDeque;
import java.util.Arrays;

public class NextGreater {
    public static void main(String[] args) {
        int[] arr = {7,3,5,12,8};

        int[] result = next_greater_element(arr);
        System.out.println(Arrays.toString(result));
    }

    public static int[] next_greater_element(int[] arr){
        Deque<Character> stack = new ArrayDeque<>();
        int n = arr.length;
        int[] result = new int[n];
        Arrays.fill(result, -1);

        for(int i=n-1;i>=0;i--){
            int current = arr[i];
            while (!stack.isEmpty() && stack.peek()<=current) {
                stack.pop();
            }
            if (!stack.isEmpty()) {
                result[i] = stack.peek();
            }
            stack.push((char) current);
        }
        return result;
    }
}

File: BasicToAdvance/Python/Day-10/NextGreater.py
arr = [12,10,14,7,15]

def next_greater_element(arr):
    result = [-1]*len(arr)
    stack = []

    for i in range(len(arr) - 1, -1, -1):

        current = arr[i]
        while stack and stack[-1]<=current:
            stack.pop()

        if stack:
            result[i] = stack[-1]

        stack.append(current)
    return result

res = next_greater_element(arr)
print(arr)
print(res)